src/i18n.ts

```diff
@@ -0,0 +1,316 @@
+  }
+
+  /** Returns a middleware to .use on the `Bot` instance. */
+  middleware(): MiddlewareFn<C & I18nFlavor> {
+    return middleware(this.fluent, this.config);
+  }
+}
+
+function middleware<C extends Context = Context>(
+  fluent: Fluent,
+  { defaultLocale, localeNegotiator, useSession }: I18nConfig<C>,
+) {
+  return async function (
+    ctx: C & I18nFlavor,
+    next: NextFunction,
+  ): Promise<void> {
+    let translate: TranslateFunction;
+
+    function useLocale(locale: LocaleId): void {
+      translate = fluent.withLocale(locale);
+    }
+
+    async function getNegotiatedLocale(): Promise<string> {
+      return await localeNegotiator?.(ctx) ??
+        // deno-lint-ignore no-explicit-any
+        (await (useSession && (ctx as any).session))?.__language_code ??
+        ctx.from?.language_code ??
+        defaultLocale;
+    }
+
+    async function setLocale(locale: LocaleId): Promise<void> {
+      if (!useSession) {
+        throw new Error(
+"You are calling `ctx.i18n.setLocale()` without setting `useSession` to `true` \
+in the configuration. It doesn't make sense because you cannot set a locale in \
+the session that way. When you call `ctx.i18n.setLocale()`, the bot tries to \
+store the user locale in the session storage. But since you don't have session \
+enabled, it cannot store the locale information in the session storage. You \
+should either enable sessions or use `ctx.i18n.useLocale()` instead.",
+        );
+      }
+
+      // deno-lint-ignore no-explicit-any
+      (await (ctx as any).session).__language_code = locale;
+      await negotiateLocale();
+    }
+
+    // Determining the locale to use for translations
+    async function negotiateLocale(): Promise<void> {
+      const negotiatedLocale = await getNegotiatedLocale();
+      useLocale(negotiatedLocale);
+    }
+
+    // Also exports ctx object properties for accessing them directly from
+    // the translation source files.
+    function translateWrapper(
+      key: string,
+      context?: TranslationContext,
+    ): string {
+      return translate(key, {
+        first_name: ctx.from?.first_name ?? "",
+        ctx: JSON.stringify(makeContextObject(ctx)),
+        ...context,
+      });
+    }
+
+    ctx.i18n = {
+      fluent,
+      renegotiateLocale: negotiateLocale,
+      useLocale,
+      getLocale: getNegotiatedLocale,
+      setLocale,
+    };
+    ctx.t = translateWrapper;
+    ctx.translate = translateWrapper;
+
+    await negotiateLocale();
+    await next();
+  };
+}
+
+/**
+ * A filter middleware for listening to the messages send by the in their language.
+ * It is useful when you have to listen for custom keyboard texts.
+ *
+ * ```ts
+ * bot.filter(hears("menu-btn"), (ctx) => ...)
+ * ```
+ *
+ * @param key Key of the message to listen for.
+ */
+export function hears(key: string) {
+  return function <C extends Context & I18nFlavor>(ctx: C) {
+    const msg = ctx.message ?? ctx.channelPost;
```

Comment:
`ctx.msg` maybe?

Reply:
Nice catch! Thanks. I'll change it. I copied the thing as it was `Composer::hears` from grammY 1.9.x iirc.

Reply:
This is a bad idea. ctx.msg gets the message from many more places. For example, it would cause hears to trigger for callback queries and edited messages, which is undesired. This code is correct.

Reply:
Right. I think, we should refactor it to use `Context.has.text()` as well

Reply:
Yep, that's a great idea!

Reply:
We'll lose support for older grammY versions though

Reply:
That's correct … if you wish to support older versions, you will need to duplicate the code, as there is no way to reuse this logic before 1.1

Reply:
What would you suggest? Support them or not?